app/models/opponent.rb

```diff
@@ -0,0 +1,108 @@
+class Opponent < ApplicationRecord
+  CASE_RELATIONSHIP_REGEX = /^relationship_case_(\S+)\?$/.freeze # begins with 'relationship_case_' and ends with a question mark
+  CLIENT_RELATIONSHIP_REGEX = /^relationship_client_(\S+)\?$/.freeze # begins with "relationship_client_' and ends with a question mark
+
+  CLIENT_RELATIONSHIP_VALUES = {
+    civil_partner: 'Civil partner',
+    customer: 'Customer',
+    employee: 'Employee',
+    employer: 'Employer',
+    ex_civil_partner: 'Ex civil partner',
+    ex_husband_wife: 'Ex_spouse',
+    grandparent: 'Grandparent',
+    husband_wife: 'Husband / wife',
+    landlord: 'Landlord',
+    legal_guardian: 'Legal guardian',
+    local_authority: 'Local authority',
+    medical_professional: 'Medial professional',
+    none: 'None',
+    other_family_member: 'Other family members',
+    other_party_type: 'Other party type',
+    parent: 'Parent',
+    property_owner: 'Property owner',
+    solicitor_barrister: 'Solicitor / barrister',
+    step_parent: 'Step parent',
+    supplier: 'Supplier',
+    tenant: 'Tenant'
+  }.freeze
+
+  CASE_RELATIONSHIP_VALUES = {
+    agent: 'Agent',
+    child: 'Child',
+    intervenor: 'Intevenor',
+    opponent: 'Opp',
+    guardian: 'Guardian',
+    beneficiary: 'Beneficiary',
+    interested_party: 'Interested party'
+  }.freeze
+
+  def shared_ind
+    false
+  end
+
+  def assessed_income
+    0
+  end
+
+  def assessed_assets
+    0
+  end
+
+  def full_name
+    "#{title.capitalize} #{first_name} #{surname}"
+  end
+
+  def person?
+    opponent_type.capitalize == 'Person'
+  end
+
+  def organisation?
+    opponent_type.capitalize == 'Organisation'
+  end
+
+  def other_party_relationship_to_client
+    relationship_to_client.capitalize
+  end
+
+  def other_party_relationship_to_case
+    relationship_to_case.capitalize
+  end
+
+  def method_missing(meth, *args)
+    if valid_relationship_to_case_method?(meth)
+      case_relationship?(meth)
+    elsif valid_relationship_to_client_method?(meth)
+      client_relationship?(meth)
+    else
+      super
+    end
+  end
```

Comment:
Using meta-programming when there is a simpler alternative is ALWAYS a mistake. The code is over complicated, difficult to debug and extend. I strongly recommend replacing this `method_missing` functionality is replaced with:

```ruby
def case_relationship?(relationship)
  relationship_to_case.capitalize == CASE_RELATIONSHIP_VALUES[relationship.to_sym]
end

def client_relationship?(relationship)
  relationship_to_client.capitalize == CLIENT_RELATIONSHIP_VALUES[relationship.to_sym]
end
```
And thus:
```
# rather than:
opponent.relationship_case_child?

# We do this:
opponent.relationship_case? :child
```

Reply:
OK - I understand the requirement for this at the moment.